# nf-core/yascp: Output

## Introduction

This guide provides a comprehensive overview of the outputs generated by the pipeline.

The pipeline will create the following files in your working directory:

```console
work            # Directory containing the nextflow working files
results         # Finished results (configurable, see below)
.nextflow_log   # Log file from Nextflow
# Other nextflow hidden files, eg. history of pipeline runs and old logs.
```

## Pipeline Overview

Utilizing [Nextflow](https://www.nextflow.io/), our pipeline orchestrates a series of data processing steps. The structure of the overall results folder is outlined below, offering a snapshot of the diverse outputs from different stages of the pipeline:

- **[preprocessing](#preprocessing)**: ???
- **[doublet_detection](#doublet_detection)**: A folder with identified artificial doublet cells.
- **[deconvolution](#deconvolution)**: A folder with the results of disentangling mixtures of cells from different donors.
- **[celltype_assignemt](#celltype_assignemt)**: A folder with cells classified into types.
- **[clustering_and_integration](#clustering_and_integration)**:  A folder with cells grouped based on similarities and integrating datasets.
- **[citeseq](#citeseq)**: A folder with CITE-seq (Cellular Indexing of Transcriptomes and Epitopes by Sequencing) data.
- **[handover](#handover)**: ???
- **[pipeline_info]**: ???
- **[yascp_inputs]**: ???

Detailed explanations of each output folder and their corresponding steps are provided below:

## preprocessing
```
preprocessing/
├── data_modalities_split
│   ├── filterd
│   │   └── Pool1
│   │       ├── Gene_Expression-Pool1.h5ad
│   │       └── Pool1__Gene_Expression
│   ├── preprocess
│   │   └── Pool1
│   │       ├── Gene_Expression-Pool1.h5ad
│   │       └── Pool1__Gene_Expression
│   └── raw
│       └── Pool1
│           ├── Gene_Expression-Pool1.h5ad
│           └── Pool1__Gene_Expression
├── recourses
│   ├── Done.tmp
│   ├── full_test_dataset
│   ├── input_test_data_file.tsv
│   └── input_test_vcf_file.tsv
└── subset_genotypes
    ├── Genotype___AllExpectedGT_Pool1
    └── Genotypes_all_pools.tsv

```

The folder preprocessing contains the next folders
- **[data_modalities_split]
- **[recourses]
- **[subset_genotypes]

## doublet_detection
```
doublet_detection
├── DoubletDecon
│   └── Pool1__DoubletDecon_doublets_singlets.tsv
├── DoubletFinder
│   └── Pool1__DoubletFinder_doublets_singlets.tsv
├── doublet_results_combined
│   └── Pool1__doublet_results_combined.tsv
├── droplet_type_distribution
│   └── Pool1__droplet_type_distribution.png
├── scDblFinder
│   └── Pool1__scDblFinder_doublets_singlets.tsv
├── SCDS
│   └── Pool1__scds_doublets_singlets.tsv
└── scrublet
    ├── plots
    │   ├── Pool1boxplot_total_umi_counts.png
    │   ├── Pool1histogram_multiplet_scores_log.png
    │   ├── Pool1histogram_multiplet_scores.png
    │   └── Pool1histogram_multiplet_zscores.png
    └── Pool1scrublet.tsv
```
DoubletFinder, DoubletDecon, scDblFinder, SCDS contain tsv files per pool with a barcode and label of whether it's a singlet or a doublet
scrublet contains tsv files per pool with a barcode and label of whether it's a multiplet or not and a folder with plots
doublet_results_combined contains tsv files per pool with barcode and labels from scrublet,	scds, scDblFinder, DoubletDecon, DoubletFinder
droplet_type_distribution contains png files with graphs showing ...

## deconvolution

- **[vireo_raw](#vireo_raw)**:
- **[vireo_processed](#vireo_processed)**:
- **[vireo_sub](#vireo_sub)**:
- **[infered_genotypes](#infered_genotypes)**:
- **[split_donor_h5ad](#split_donor_h5ad)**:
- **[filepaths](#filepaths)**:
- **[cellsnp](#cellsnp)**:
- **[existing_cellsnp](#existing_cellsnp)**: (name can be different)
- **[concordances](#concordances)**:
- **[deconvolution_results](#deconvolution_results)**:
- **[gtmatch](#gtmatch)**:
infered_genotypes



## vireo_raw
```
vireo_raw
├── correlations.png
├── donor_corelations_matrix.tsv
├── matched_donors.txt
└── Pool1
    └── vireo_Pool1

vireo_raw/
├── correlations.png
├── CRD_CMB13450877
│   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   └── vireo_CRD_CMB13450877
├── CRD_CMB13450878
│   ├── dubs_removed__Study_Merge_AllExpectedGT_F24ROJ53N_out.vcf.gz
│   ├── dubs_removed__Study_Merge_AllExpectedGT_F24ROJ53N_out.vcf.gz.csi
│   ├── sub_CRD_CMB13450878_Expected.vcf.gz
│   └── vireo_CRD_CMB13450878
├── donor_corelations_matrix.tsv
└── matched_donors.txt

```

## vireo_processed
```
vireo_processed
├── assignments_all_pools.tsv
└── Pool1
    ├── GT_replace_donor_ids_false.tsv
    ├── GT_replace_GT_donors.vireo_false.vcf.gz
    ├── GT_replace_Pool1_assignments_false.tsv
    ├── GT_replace_Pool1__exp.sample_summary_false.txt
    └── GT_replace_Pool1.sample_summary_false.txt

vireo_processed/
├── assignments_all_pools.tsv
├── CRD_CMB13450877
│   ├── GT_replace_CRD_CMB13450877_assignments_true.tsv
│   ├── GT_replace_CRD_CMB13450877__exp.sample_summary_true.txt
│   ├── GT_replace_CRD_CMB13450877.sample_summary_true.txt
│   ├── GT_replace_donor_ids_true.tsv
│   └── GT_replace_GT_donors.vireo_true.vcf.gz
└── CRD_CMB13450878
    ├── GT_replace_CRD_CMB13450878_assignments_true.tsv
    ├── GT_replace_CRD_CMB13450878__exp.sample_summary_true.txt
    ├── GT_replace_CRD_CMB13450878.sample_summary_true.txt
    ├── GT_replace_donor_ids_true.tsv
    └── GT_replace_GT_donors.vireo_true.vcf.gz

```
## vireo_sub
```
vireo_sub
├── CRD_CMB13450877
│   ├── vireo_____1
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   │   └── vireo_CRD_CMB13450877___1
│   ├── vireo_____10
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   │   └── vireo_CRD_CMB13450877___10
│   ├── vireo_____2
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   │   └── vireo_CRD_CMB13450877___2
│   ├── vireo_____3
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   │   └── vireo_CRD_CMB13450877___3
│   ├── vireo_____4
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   │   └── vireo_CRD_CMB13450877___4
│   ├── vireo_____5
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   │   └── vireo_CRD_CMB13450877___5
│   ├── vireo_____6
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   │   └── vireo_CRD_CMB13450877___6
│   ├── vireo_____7
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   │   └── vireo_CRD_CMB13450877___7
│   ├── vireo_____8
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   │   └── vireo_CRD_CMB13450877___8
│   └── vireo_____9
│       ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│       ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│       ├── sub_CRD_CMB13450877_Expected.vcf.gz
│       └── vireo_CRD_CMB13450877___9
└── CRD_CMB13450878
    ├── vireo_____1
    │   ├── dubs_removed__Study_Merge_AllExpectedGT_F24ROJ53N_out.vcf.gz
```
## infered_genotypes
```
infered_genotypes
├── CRD_CMB13450877
│   ├── CRD_CMB13450877_headfix_vireo.vcf.gz
│   └── CRD_CMB13450877_headfix_vireo.vcf.gz.tbi
└── CRD_CMB13450878
    ├── CRD_CMB13450878_headfix_vireo.vcf.gz
    └── CRD_CMB13450878_headfix_vireo.vcf.gz.tbi
```

## split_donor_h5ad
```
split_donor_h5ad
└── Pool1
    ├── cell_belongings.tsv
    ├── donor_level_anndata
    │   ├── donor0.Pool1.barcodes.tsv
    │   ├── donor0.Pool1.h5ad
    │   ├── donor1.Pool1.barcodes.tsv
    │   ├── donor1.Pool1.h5ad
    │   ├── donor2.Pool1.barcodes.tsv
    │   ├── donor2.Pool1.h5ad
    │   ├── doublet.Pool1.barcodes.tsv
    │   ├── doublet.Pool1.h5ad
    │   ├── unassigned.Pool1.barcodes.tsv
    │   └── unassigned.Pool1.h5ad
    ├── Pool1.donors.h5ad.assigned.tsv
    ├── Pool1__donors.h5ad.assigned.tsv
    ├── Pool1.donors.h5ad.tsv
    ├── Pool1__donors.h5ad.tsv
    ├── Pool1_exp__donor_n_cells.tsv
    ├── Pool1.h5ad.tsv
    ├── vireo_annot.Pool1.h5ad
    └── Vireo_plots.pdf
```

## filepaths
```
filepaths
├── cellranger_as_h5ad.tsv
├── donors_h5ad_assigned.tsv
├── donors_h5ad.tsv
├── exp__donors_h5ad_assigned.tsv
├── exp__donors_h5ad.tsv
├── vireo_donor_n_cells.tsv
└── vireo_exp__donor_n_cells.tsv
```

## cellsnp
```
cellsnp
└── cellsnp_Pool1
    ├── cellSNP.base.vcf.gz
    ├── cellSNP.cells.vcf.gz
    ├── cellSNP.samples.tsv
    ├── cellSNP.tag.AD.mtx
    ├── cellSNP.tag.DP.mtx
    └── cellSNP.tag.OTH.mtx
```

## existing_cellsnp

## concordances
```
concordances
└── all_variants_description.tsv

concordances/
├── all_variants_description.tsv
├── becoming_different_donor.png
├── becoming_doublet_donor.png
├── becoming_unassigned_donor.png
├── CRD_CMB13450877
│   ├── 1090095_1090095-donor3--each_cells_comparison_with_other_donor.tsv
│   ├── 1709635_1709635-donor5--each_cells_comparison_with_other_donor.tsv
│   ├── 2288590_2288590-donor6--each_cells_comparison_with_other_donor.tsv
│   ├── 2743244_2743244-donor7--each_cells_comparison_with_other_donor.tsv
│   ├── 2768849_2768849-donor4--each_cells_comparison_with_other_donor.tsv
│   ├── 2998395_2998395-donor2--each_cells_comparison_with_other_donor.tsv
│   ├── 3183427_3183427-donor0--each_cells_comparison_with_other_donor.tsv
│   ├── 3699286_3699286-donor1--each_cells_comparison_with_other_donor.tsv
│   ├── 4853673_4853673-donor9--each_cells_comparison_with_other_donor.tsv
│   ├── 5154993_5154993-donor8--each_cells_comparison_with_other_donor.tsv
│   ├── becoming_different_donor.png
│   ├── becoming_doublet_donor.png
│   ├── becoming_unassigned_donor.png
│   ├── cell_belongings.tsv
│   ├── cellSNP.cells.vcf.gz
│   ├── CRD_CMB13450877__joined_df_for_plots.tsv
│   ├── CRD_CMB13450877_subsampling_donor_swap_quantification.tsv
│   ├── Discordant_reads_becoming_different_donor_no0.png
│   ├── Discordant_reads_becoming_different_donor.png
│   ├── Discordant_reads_by_n_sites_becoming_different_donor_no0.png
│   ├── Discordant_reads_by_n_sites_becoming_different_donor.png
│   ├── discordant_sites_in_other_donors_noA2G.tsv
│   ├── Nr_discordant_uninformative_becoming_different_donor.png
│   ├── sites_becoming_different_donor_no0.png
│   ├── sites_becoming_different_donor.png
│   ├── sites_becoming_different_donor_probs.png
│   ├── sites_becoming_doublet_donor.png
│   ├── sites_becoming_unassigned_donor.png
│   ├── sites_vs_concordance.png
│   ├── stats_CRD_CMB13450877_gt_donor_assignments.csv
│   ├── sub_CRD_CMB13450877_Expected.vcf.gz
│   ├── sub_CRD_CMB13450877_GT_Matched.vcf.gz
│   ├── subplot_sites_vs_concordance.png
│   ├── tmp_donor_distinct_sites.pkl
│   ├── tmp_exclusive_cell_variants.pkl
│   ├── tmp_exclusive_don_variants.pkl
│   ├── tmp_GT_Expected_variants.pkl
│   ├── tmp_GT_Matched_variants.pkl
│   └── Total_reads_becoming_different_donor.png -> ../../../../work/bb/e8336817b1df57e119d24a67aa5dcb/Total_reads_becoming_different_donor.png
├── Discordant_reads_becoming_different_donor_no0.png
├── Discordant_reads_becoming_different_donor.png
├── Discordant_reads_by_n_sites_becoming_different_donor_no0.png
├── Discordant_reads_by_n_sites_becoming_different_donor.png
├── joined_df_for_plots.tsv
├── Nr_discordant_uninformative_becoming_different_donor.png
├── sites_becoming_different_donor_no0.png
├── sites_becoming_different_donor.png
├── sites_becoming_different_donor_probs.png
├── sites_becoming_doublet_donor.png
├── sites_becoming_unassigned_donor.png
├── sites_vs_concordance.png
├── subplot_sites_vs_concordance.png
└── Total_reads_becoming_different_donor.png

```

## infered_genotypes

```
infered_genotypes
└── Pool1
    ├── Pool1_headfix_vireo.vcf.gz
    └── Pool1_headfix_vireo.vcf.gz.tbi
```

## gtmatch
```
gtmatch/
├── assignments_all_pools.tsv
└── Pool1
    ├── Done.tmp
    ├── Expected_Withing_expected_Pool1.genome
    ├── GT_replace_PiHAT_Stats_File_Pool1.csv
    ├── InferedExpected_Expected_Infered_Pool1.genome
    ├── InferedGTMatched_Expected_Infered_Pool1.genome
    ├── InferedOnly_Withing_pool_Pool1.genome
    ├── PiHAT_Stats_File_Pool1.csv
    ├── Pool1_gt_donor_assignments.csv
    ├── pool_Pool1_panel_Pool1_Onek1K_gtcheck_donor_assignments.csv
    ├── pool_Pool1_panel_Pool1_Onek1K_gtcheck_score_table.csv
    └── stats_Pool1_gt_donor_assignments.csv
```


## celltype_assignemt
```
celltype_assignemt/
├── All_Celltype_Assignments.tsv
├── azimuth
│   └── PBMC
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.mapping_score_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.mapping_score_vln.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.ncells_by_type_barplot.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.prediction_score_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.prediction_score_vln.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.query_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l2.mapping_score_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l2.mapping_score_vln.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l2.ncells_by_type_barplot.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l2.prediction_score_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l2.prediction_score_vln.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l2.query_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l3.mapping_score_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l3.mapping_score_vln.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l3.ncells_by_type_barplot.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l3.prediction_score_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l3.prediction_score_vln.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l3.query_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_predicted_celltype_l1.tsv
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_predicted_celltype_l2.tsv
│       └── AZ_1.pre_QC_adata_Pool1_Pool1_predicted_celltype_l3.tsv
├── celltypist
│   ├── COVID19_Immune_Landscape
│   │   └── Pool1
│   │       ├── Pool1___COVID19_Immune_Landscape___decision_matrix.csv
│   │       ├── Pool1___COVID19_Immune_Landscape___predicted_labels.csv
│   │       ├── Pool1___COVID19_Immune_Landscape___probability_matrix.csv
│   │       ├── Pool1_majority_voting.pdf
│   │       ├── Pool1_over_clustering.pdf
│   │       └── Pool1_predicted_labels.pdf
│   ├── Immune_All_High
│   │   └── Pool1
│   │       ├── Pool1___Immune_All_High___decision_matrix.csv
│   │       ├── Pool1___Immune_All_High___predicted_labels.csv
│   │       ├── Pool1___Immune_All_High___probability_matrix.csv
│   │       ├── Pool1_majority_voting.pdf
│   │       ├── Pool1_over_clustering.pdf
│   │       └── Pool1_predicted_labels.pdf
│   └── Immune_All_Low
│       └── Pool1
│           ├── Pool1___Immune_All_Low___decision_matrix.csv
│           ├── Pool1___Immune_All_Low___predicted_labels.csv
│           ├── Pool1___Immune_All_Low___probability_matrix.csv
│           ├── Pool1_majority_voting.pdf
│           ├── Pool1_over_clustering.pdf
│           └── Pool1_predicted_labels.pdf
├── donor_celltype_report.tsv
├── scpred
│   ├── AZ_1.pre_QC_adata_Pool1_AZ_1.pre_QC_adata_Pool1__scpred_prediction.tsv
│   └── AZ_1.pre_QC_adata_Pool1_hier_scpred.RDS
└── tranche_celltype_report.tsv
```
- **[All_Celltype_Assignments.tsv]
- **[donor_celltype_report.tsv]
- **[tranche_celltype_report.tsv]
- **[scored]
- **[azimuth]
- **[celltypist]
- Immune_All_Low
- COVID19_Immune_Landscape
- Immune_All_High

## clustering_and_integration
```
clustering_and_integration/
├── normalize=total_count.vars_to_regress=none
│   ├── adatametadata.tsv.gz
│   ├── adatanormalized.h5ad
│   ├── adatanormalized_pcacounts.h5ad
│   ├── adatanormalized_pca.h5ad
│   ├── adatanormalized_pcaknee.tsv
│   ├── adatapcs.tsv.gz
│   ├── donor_level_anndata_QCfiltered
│   │   └── Pool1___sample_QCd_adata.h5ad
│   ├── plots
│   │   ├── adatanormalized_pcakneevariance_ratiospline=interp1dknee_normalized.png
│   │   ├── adatanormalized_pcakneevariance_ratiospline=interp1dknee_raw.png
│   │   ├── adatanormalized_pcakneevariance_ratiospline=Noneknee_normalized.png
│   │   ├── adatanormalized_pcakneevariance_ratiospline=Noneknee_raw.png
│   │   ├── adatanormalized_pcakneevariancespline=interp1dknee_normalized.png
│   │   ├── adatanormalized_pcakneevariancespline=interp1dknee_raw.png
│   │   ├── adatanormalized_pcakneevariancespline=Noneknee_normalized.png
│   │   ├── adatanormalized_pcakneevariancespline=Noneknee_raw.png
│   │   ├── filter_genes_dispersionadata.pdf
│   │   ├── highest_expr_genesadata.pdf
│   │   ├── pca_variance_ratioadatalog.pdf
│   │   └── pca_variance_ratioadata.pdf
│   └── reduced_dims-null-pca.n_pcs=20
│       ├── clustering_and_integration
│       │   └── plots
│       │       ├── pca_loadings-pca-n_pcs=20.png
│       │       ├── pca-pca-Azimuth:predicted.celltype.l2.score.png
│       │       ├── pca-pca-experiment_id.png
│       │       ├── pca-pca-log10_ngenes_by_count.png
│       │       ├── pca-pca-n_cells.png
│       │       ├── pca-pca-pct_counts_gene_group__mito_transcript.png
│       │       ├── pca-pca-pct_counts_gene_group__ribo_rna.png
│       │       ├── pca-pca-prob_doublet.png
│       │       └── pca-pca-total_counts.png
│       └── reduced_dims.tsv.gz
└── plots
    ├── mads-n_genes_by_counts.png
    ├── mads-pct_counts_gene_group__mito_protein.png
    ├── mads-pct_counts_gene_group__mito_transcript.png
    ├── mads-pct_counts_gene_group__ribo_protein.png
    ├── mads-pct_counts_gene_group__ribo_rna.png
    ├── mads-total_counts.png
    ├── mads.tsv
    ├── plot_ecdf.var=pct_counts_gene_group__mito_transcript.color=experiment_id-outfile.png
    ├── plot_ecdf.var=total_counts.color=experiment_id-outfile.png
    ├── plot_ecdf-x_log10.var=pct_counts_gene_group__mito_transcript.color=experiment_id-outfile.png
    ├── plot_ecdf-x_log10.var=total_counts.color=experiment_id-outfile.png
    ├── plot_histogram.var=pct_counts_gene_group__mito_transcript.facet=experiment_id-outfile.png
    ├── plot_histogram.var=total_counts.facet=experiment_id-outfile.png
    ├── plot_histogram-x_log10.var=pct_counts_gene_group__mito_transcript.facet=experiment_id-outfile.png
    ├── plot_histogram-x_log10.var=total_counts.facet=experiment_id-outfile.png
    ├── plot_nfeature_mt_cellpassqc.facet=experiment_id-outfile.png
    ├── plot_nfeature_mt_density.facet=experiment_id-outfile.png
    ├── plot_umi_mt_cellpassqc.facet=experiment_id-outfile.png
    ├── plot_umi_mt_density.facet=experiment_id-outfile.png
    ├── plot_umi_ngene_cellpassqc.facet=experiment_id-outfile.png
    ├── plot_umi_ngene_mt_density.facet=experiment_id-outfile.png
    ├── plot_umi_ngene_mt.facet=experiment_id-outfile.png
    └── scatterplot-sex_sample_swap_check.png


clustering_and_integration/
├── normalize=total_count.vars_to_regress=none
│   ├── adatametadata.tsv.gz
│   ├── adatanormalized.h5ad
│   ├── adatanormalized_pcacounts.h5ad
│   ├── adatanormalized_pca.h5ad
│   ├── adatanormalized_pcaknee.tsv
│   ├── adatapcs.tsv.gz
│   ├── donor_level_anndata_QCfiltered
│   │   └── Pool1___sample_QCd_adata.h5ad
│   ├── plots
│   ├── reduced_dims-null-bbknn.batch=experiment_id.n_pcs=20
│   │   ├── cluster.number_neighbors=-1.method=leiden.resolution=0.1
│   │   │   ├── clustering_0.1clustered.h5ad
│   │   │   ├── clustering_0.1clustered.tsv.gz
│   │   │   ├── dotplot_sampleclustering_0.1clustered_ncells0.pdf
│   │   │   ├── dotplot_sampleclustering_0.1clustered_ncellsless5.pdf
│   │   │   ├── dotplot_sampleclustering_0.1clustered.pdf
│   │   │   ├── plots
│   │   │   ├── sccaf
│   │   │   └── validate_resolution
│   │   ├── cluster.number_neighbors=-1.method=leiden.resolution=0.5
│   │   ├── cluster.number_neighbors=-1.method=leiden.resolution=1.0
│   │   ├── cluster.number_neighbors=-1.method=leiden.resolution=5.0
│   │   ├── outfile_adatabbknn.h5ad
│   │   ├── plots
│   │   ├── reduced_dims.tsv.gz
│   │   ├── resolution_tuningmerged_model_report.tsv.gz
│   │   ├── resolution_tuningmerged_test_result.tsv.gz
│   │   └── umap_gather_out.h5ad
│   ├── reduced_dims-null-harmony.n_pcs=20.variables=experiment_id.thetas=1.0
│   │   ├── cluster.number_neighbors=15.method=leiden.resolution=0.1
│   │   ├── cluster.number_neighbors=15.method=leiden.resolution=0.5
│   │   ├── cluster.number_neighbors=15.method=leiden.resolution=1.0
│   │   ├── cluster.number_neighbors=15.method=leiden.resolution=5.0
│   │   ├── plots
│   │   ├── reduced_dims.tsv.gz
│   │   ├── resolution_tuningmerged_model_report.tsv.gz
│   │   ├── resolution_tuningmerged_test_result.tsv.gz
│   │   └── umap_gather_out.h5ad
│   └── reduced_dims-null-pca.n_pcs=20
│       ├── clustering_and_integration
│       │   └── plots
│       └── reduced_dims.tsv.gz
└── plots

```


normalize=total_count.vars_to_regress=none (should it be renamed)
plots

## citeseq
```
citeseq/
└── DSB
    └── Pool1
        └── CITE__Pool1
```
- **[DSB] is there only one folder in citeseq? DSB has folders for each pool

## handover
- **[Summary_plots]
- **[Donor_Quantification]
- **[Donor_Quantification_summary]
- **[merged_h5ad]

```
handover/
├── Donor_Quantification
├── Donor_Quantification_summary
├── merged_h5ad
└── Summary_plots
└── UMAPs
```

## Pipeline Info

## YASCP Inputs

